### PR TITLE
Fix typescript typings for Storage

### PR DESCRIPTION
**Pull request by cliffkoh**
This fix is per the discussion at https://github.com/Microsoft/TypeScript/issues/18397. I'm positive this typing is the actual intent of the generic `Result` :)

### CODE REVIEW

File: type-definitions/index.d.ts [hunk 1]
```diff
@@ -1,19 +1,18 @@
...
     ): Promise<Result | string>;
     removeItem(key: string, onComplete?: OnComplete<any>): Promise<any>;
-    getAllKeys?<Result>(
+    getAllKeys?(
       callback?: (error?: Error, keys?: string[]) => void
     ): Promise<string[]>;
-    getAllKeys?<Result>(onComplete?: OnComplete<Result>): Promise<Result>;
+    getAllKeys?(onComplete?: OnComplete<Result>): Promise<Result>;
     keys?: (...args: any[]) => any;
-    [key: string]: any; // In case Storage object has some other (private?) methods and properties.
```
Comment (cliffkoh) [inline type-definitions/index.d.ts]: This is not needed for an interface - the concrete Storage object would be of the right type (e.g. AsyncStorage) and able to access those other methods and properties, while redux-persist does not care for the existence of those other methods and properties for it only relies on the methods/properties defined here in the interface.

**Comment by cliffkoh** [comment]
I took a closer look at the type definitions and noticed that there was a bit of a hack added to make it work with AsyncStorage (I was using an older version where I also hit the same problem as @iRoachie). I cleaned that up and also make the signatures compatible with some of the variants of StorageAPIs.

@iRoachie it would be great if you can review the change. I tested it and it works fine with AsyncStorage for me (pass in String to the generic).

**Comment by iRoachie** [comment]
Good job mate, I was meaning to do it at the time. Just by looking at it, it looks good

**Comment by rt2zz** [comment]
I am not a TS user, but this looks good to me. If anyone finds issue please open.

**Comment by cliffkoh** [comment]
@rt2zz any chance you can release a new patch build to npm which picks up this change? :)

**Pull request merged**